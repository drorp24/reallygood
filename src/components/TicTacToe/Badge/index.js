import React from "react";

import styles from "./styles.module.scss";

import Trophy from "../assets/trophy.svg";
import Crystal from "../assets/crystal.svg";

const Badge = ({ text, background, color, children }) => (
  <div className={styles.badge} style={{ background, color }}>
    <span>{text}</span>
    {children}
  </div>
);

export default Badge;
