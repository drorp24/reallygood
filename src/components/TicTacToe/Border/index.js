import React from "react";

import styles from "./styles.module.scss";

const Border = ({ line = null, row = null }) => {
  const classes = [styles.Border];
  if (line !== null) {
    classes.push(styles.line);
    classes.push([styles.firstLine, styles.secondLine][line]);
  } else if (row !== null) {
    classes.push(styles.row);
    classes.push([styles.firstRow, styles.secondRow][row]);
  }

  return <div className={classes.join(" ")} />;
};

export default Border;
