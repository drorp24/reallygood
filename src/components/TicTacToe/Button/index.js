import React from "react";

import styles from "./styles.module.scss";
import lightenDarkenColor from "../../../utils/lightenDarkenColor";

// Of course the shadow color should be the button's bg color darkened rather than black
const Button = ({ children, onClick, background, shadow }) => {
  const darkerColor = lightenDarkenColor(background, -30);
  return (
    <div
      className={styles.root}
      onClick={onClick}
      style={{
        backgroundColor: background,
        boxShadow: shadow && `0px 5px ${darkerColor}`,
        color: darkerColor,
      }}
    >
      {children}
    </div>
  );
};

export default Button;
