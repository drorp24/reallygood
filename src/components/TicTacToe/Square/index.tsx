// @ts-nocheck

import React, { memo } from "react";

import styles from "./styles.module.scss";

// I sometimes assume there could be some more ('...rest') arguments that I need to pass onwards to the child
// The reason I did it here is that in a real app I would want to highlight the 'winning squares'.
// To do that I would need to control the styling of the Square component from its parent
// and '...rest' could help in that.
export default memo(function Square({ value, onClick, ...rest }): JSX.Element {
  // TODO: fix...
  const color = value === "X" ? "#50a1d2" : "#333";
  return (
    <div className={styles.Square} style={{ color }} {...{ onClick, ...rest }}>
      {value}
    </div>
  );
});
