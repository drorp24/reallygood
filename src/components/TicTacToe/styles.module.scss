.TicTacToe {
  display: grid;
  grid-template-rows: [player] 15% [board] 50% [game] 35%;
  grid-gap: 10px;
  height: 95vh;
  font-weight: 700;

  .instructions {
    margin-top: 5px;
    margin-bottom: 5px;
    font-weight: bold;
    font-size: 16px;
  }

  .button {
    margin-top: 15px;
    margin-bottom: 16px;
    width: 80px;
    height: 40px;
    background-color: rgb(138, 202, 202);
    color: white;
    font-size: 16px;
  }

  .Player {
    grid-row: player;
    display: grid;
  }

  .Board {
    grid-row: board;
    display: flex;
    align-items: center;
    justify-content: space-around;
    flex-direction: column;
    text-shadow: 1px 1px 1px rgba(80, 161, 210, 0.3);
    -webkit-text-stroke: 2px #ffffff;
  }

  .Game {
    grid-row: game;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }

  .row {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr;
    width: 100%;
    justify-content: space-around;
    height: 33%;
  }
}
