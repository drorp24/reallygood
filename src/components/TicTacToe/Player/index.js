// @ts-nocheck
import React from "react";

import styles from "./styles.module.scss";

import Captain from "../assets/captain.svg";
import Batman from "../assets/batman.svg";
import Crystal from "../assets/crystal.svg";
import Trophy from "../assets/trophy.svg";

import Badge from "../Badge";

// See comments about using theme as opposed to hard-coded colors
const Player = ({ gameOver, currentPlayer, winner }) => {
  const next = (symbol) => {
    if (winner && winner === symbol) {
      return (
        <Badge
          text="Winner"
          color="#fcbd11"
          background="rgba(251, 212, 58, 0.44)"
        >
          <img src={Trophy} alt="Trophy" />
        </Badge>
      );
    }
    if (!gameOver && currentPlayer === symbol) {
      return (
        <Badge text="Plays Next" background="white" color="#a061ff">
          <img src={Crystal} alt="Crystal" />
        </Badge>
      );
    }
    return "";
  };

  // The `${styles.symbol} x` doesnt work
  // but with css-in-js this should have been a piece of cake, without all this class concatenation
  return (
    <div className={styles.container}>
      <div className={styles.player1}>
        <div className={styles.playerBox}>
          <div className={`${styles.symbol} x`}>X</div>
          <div className={styles.playerLine}>
            <div>
              <img src={Captain} alt="Captain" />
            </div>
            <div>Player 1</div>
          </div>
          <div className={styles.next}>{next("X")}</div>
        </div>
      </div>
      <div className={styles.player2}>
        <div className={styles.playerBox}>
          <div className={`${styles.symbol} o`}>O</div>
          <div className={styles.playerLine}>
            <div>Player 2</div>
            <div>
              <img src={Batman} alt="Captain" />
            </div>
          </div>
          <div className={styles.next}>{next("O")}</div>
        </div>
      </div>
    </div>
  );
};

export default Player;
