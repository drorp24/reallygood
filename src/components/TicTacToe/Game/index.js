import React from "react";

import styles from "./styles.module.scss";
import Replay from "../assets/replay.svg";
import Star from "../assets/star.svg";
import Settings from "../assets/settings.svg";
import Button from "../Button";

// Of course the backgrounds should be taken from some theme rather than hard coded
const Game = ({ reset, gameOver, winner }) => (
  <div className={styles.gameContainer}>
    <Button onClick={reset} background="#999bfb" shadow>
      <span className="resetContainer">
        <img src={Replay} alt="Replay" />
      </span>
    </Button>
    <div style={{ visibility: winner || gameOver ? "visible" : "hidden" }}>
      <Button background="#fbc399" shadow>
        <div>Rate us</div>
        <div>
          {[...Array(5)].map((_, i) => (
            <span key={i}>
              <img src={Star} alt="Star" />
            </span>
          ))}
        </div>
      </Button>
    </div>
    <Button background="#9E9E9E" shadow>
      <div>Settings</div>
      <img src={Settings} alt="Settings" />
    </Button>
  </div>
);

export default Game;
