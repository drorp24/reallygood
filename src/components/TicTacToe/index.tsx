// @ts-nocheck
import React, { memo, useState } from "react";

import Square from "./Square";
import styles from "./styles.module.scss";
import game from "../../config/game";

// while there are things that I left out of this excercise for time constrains,
// this one seemed simple enough to not steal too much of my time (and it definitely makes the code prettier)
// However, for the time constraint I didn't generalize the styling to adapt to the dimensions.
// So, if you would change the config to any other dimension it would work - but won't look too pretty...
const { dimensions } = game;
const squares = dimensions ** 2 - 1;

// I chose to control the entire board's state here rather than letting any Sqaure control its own 'X'/'O' state
// because I need to check here after every move if game is won or over.

const initialState = {
  emptyBoard: () =>
    new Array(dimensions).fill("").map(() => new Array(dimensions).fill("")),
  currentPlayer: "X",
  moves: 0,
  gameOver: false,
  winner: "",
};

export default memo(function TicTacToe(): JSX.Element {
  // I could place all these state parts in one object but the code seems clearer when they're split
  // I'm not destructuring initialState for the same reason
  const [board, setBoard] = useState(initialState.emptyBoard());
  const [currentPlayer, setCurrentPlayer] = useState(
    initialState.currentPlayer
  );
  const [moves, setMoves] = useState(initialState.moves);
  const [gameOver, setGameOver] = useState(initialState.gameOver);
  const [winner, setWinner] = useState(initialState.winner);

  const togglePlayer = () =>
    setCurrentPlayer(currentPlayer === "X" ? "O" : "X");

  // Since we're checking this following every move, a potential win can only occur in either
  //   -- the last move's line or row (i.e., no need to check all other lines and rows), or
  //   -- any of the 2 diagonals (in case line & row happen to be placed on either)
  const winning = ({ line, row }) =>
    board[line].every((_, row) => board[line][row] === currentPlayer) ||
    board.reduce(
      (acc, _, line) => acc && board[line][row] === currentPlayer,
      true
    ) ||
    (line === row &&
      board.reduce(
        (acc, _, line) => acc && board[line][line] === currentPlayer,
        true
      )) ||
    (line + row === dimensions - 1 &&
      board.reduce(
        (acc, _, line) =>
          acc && board[line][dimensions - line - 1] === currentPlayer,
        true
      ));

  // By referring to the 'board' variable from the closure, handleClick can be passed to Square components
  // and is guaranteed to get board's most current value
  const handleClick = ({ line, row }) => () => {
    if (!gameOver && board[line][row] === "") {
      board[line][row] = currentPlayer;
      setBoard(board);
      setMoves(moves + 1);
      if (winning({ line, row })) {
        setWinner(currentPlayer);
        setGameOver(true);
      } else if (moves === squares) {
        setGameOver(true);
      } else {
        togglePlayer();
      }
    }
  };

  const reset = () => {
    // See comments about splitting state and not destructuring initialState
    setBoard(initialState.emptyBoard());
    setCurrentPlayer(initialState.currentPlayer);
    setMoves(initialState.moves);
    setGameOver(initialState.gameOver);
    setWinner(initialState.winner);
  };

  const Row = ({ item, line }) => (
    <div className={styles.row}>
      {item.map((_, row) => (
        <Square
          key={`${line}${row}`}
          value={board[line][row]}
          onClick={handleClick({ line, row })}
        />
      ))}
    </div>
  );

  const Board = () =>
    board.map((item, line) => {
      return <Row key={line} {...{ item, line }} />;
    });

  return (
    <div className={styles.TicTacToe}>
      <div className={styles.Container}>
        <div className={styles.instructions}>
          Next player: {gameOver ? "Game Over" : currentPlayer}
        </div>
        <div className={styles.instructions}>
          {/* I know you wanted to write nothing if no one wins */}
          Winner: {gameOver && !winner ? "Tie!" : winner}
        </div>
        <button className={styles.button} type="button" onClick={reset}>
          Reset
        </button>

        <div className={styles.Board}>
          <Board />
        </div>
      </div>
    </div>
  );
});
