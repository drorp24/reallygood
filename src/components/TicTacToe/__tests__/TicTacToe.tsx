import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import TicTacToe from "../index";

describe("<TicTacToe />", () => {
  it("renders", () => {
    const tree = render(
      <MemoryRouter>
        <TicTacToe />
      </MemoryRouter>
    ).container.firstChild;

    expect(tree).toMatchSnapshot();
  });
});
