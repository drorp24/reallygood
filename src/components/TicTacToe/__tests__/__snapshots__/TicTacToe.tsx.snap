// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicTacToe /> renders 1`] = `
<div
  class="TicTacToe"
>
  <div
    class="Container"
  >
    <div
      class="instructions"
    >
      Next player: 
      X
    </div>
    <div
      class="instructions"
    >
      Winner: 
      
    </div>
    <button
      class="button"
      type="button"
    >
      Reset
    </button>
    <div
      class="Board"
    >
      <div
        class="row"
      >
        <div
          class="Square"
        />
        <div
          class="Square"
        />
        <div
          class="Square"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="Square"
        />
        <div
          class="Square"
        />
        <div
          class="Square"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="Square"
        />
        <div
          class="Square"
        />
        <div
          class="Square"
        />
      </div>
    </div>
  </div>
</div>
`;
